# Face Mask Detection

## Introduction

In this notebook, I'll show you how to create a custom dataset for both YOLO-based object detection and Vision Transformer (ViT) Transformers-based face mask classification using the face mask detection dataset from Kaggle. This dataset contains images of people with, without masks, and mask weared incorrect, along with annotations specifying the location of faces and labels indicating mask-wearing status.

We'll explore two different approaches to utilize this dataset: one for YOLO-based object detection and another for ViT Transformers-based face mask classification. YOLO-based object detection will enable us to detect and localize faces with masks, while ViT Transformers will allow us to classify whether a person is wearing a mask or not.

## Setup

First, make sure you have all the necessary libraries installed. These include; 
- pandas
- matplotlib==3.8.4
- scikit-learn==1.4.2
- pyyaml==6.0.1
- pillow==10.3.0
- ultralytics==8.0.208
- transformers==4.40.1
- datasets==2.19.1

Here to install the libraries,
```
!pip install pandas scikit-learn==1.4.2 pyyaml==6.0.1 pillow==10.3.0 ultralytics==8.0.208 transformers==4.40.1 datasets==2.19.1

```

In [1]:
import os
import glob
import shutil
import random
import xml.etree.ElementTree as ET
import yaml
from PIL import Image
from sklearn.model_selection import train_test_split
import numpy as np
import pandas as pd

## Dataset Preparation

1. Download the face mask detection dataset from [Kaggle](https://www.kaggle.com/datasets/andrewmvd/face-mask-detection/).
2. Extract the dataset files into a directory structure containing images and annotations.

### Download face mask detection dataset

```
!wget -o face_mask_dataset.zip https://www.kaggle.com/datasets/andrewmvd/face-mask-detection/download?datasetVersionNumber=1
```

### Extract downloaded dataset

```
unzip face_mask_dataset.zip -d face_mask_dataset
```

### Dataset Generator

Here's breakdown of the process:
1. Dataset Loading and Preparation:
- The code loads image files and corresponding annotations from the dataset directory.
- It extracts image paths and annotation paths from the specified directory structure.
- The dataset is organized into images and annotations, which are used for further processing.
2. Data Transformation and Annotation Handling:
- The code processes annotations to extract bounding box coordinates and class labels for each object (e.g., face with or without a mask).
- Image annotations are converted into YOLO format, which includes class labels and bounding box coordinates normalized to image dimensions.
3.Dataset Splitting:
- The dataset is split into training, validation, and test sets using the train_test_split.
- The splitting process ensures that images and corresponding annotations are kept together in their respective sets.
4. Dataset Saving:
- The processed dataset is saved in a directory structure suitable for YOLO-based object detection.
- Images are saved in the images directory, and annotations are saved in the labels directory, with corresponding file names.
5. Configuration File Creation:
- A configuration file is generated to specify the paths to the train, validation, and test sets, along with other parameters required for model training.
- This configuration file is saved in YAML format and contains essential information for training the YOLO model.

In [2]:
class DatasetGenerator:
    def __init__(self, data_dir: str):
        self.data_dir = data_dir
        self.images_path = self._load_images()
        self.annotations_path = self._load_annotations()
        self.labels_dict = {
            "with_mask": 0,
            "without_mask": 1,
            "mask_weared_incorrect": 2,
        }
        self.images, self.labels = self._extract_dataset_(self.annotations_path)

    def __len__(self) -> int:
        return len(self.labels)

    def __getitem__(self, index: int) -> tuple:
        image = self.images[index]
        label = self.labels[index]
        return image, label

    def to_dict(self) -> dict:
        return dict(image=self.images, labels=self.labels)

    def _extract_dataset_(self, annotation_path: list[str])->tuple[list, list]:
        images = []
        labels = []
        for annotation in annotation_path:
            data = self._get_annotations(annotation)
            image = Image.open(data["image_path"]).convert("RGB")
            for bbox, label in zip(data["bbox"], data["labels"]):
                roi = image.crop(bbox)
                images.append(roi)
                labels.append(label)

        return images, labels

    def _load_images(self)->list[str]:
        return glob.glob(os.path.join(self.data_dir, "images", "*.*"))

    def _load_annotations(self)->list[str]:
        return glob.glob(os.path.join(self.data_dir, "annotations", "*.xml"))

    def _convert_coordinates(self, size: tuple, box: tuple) -> list[float]:
        xmin, ymin, xmax, ymax = box
        dw, dh = 1.0 / size[0], 1.0 / size[1]
        x = xmin
        y = ymin
        w = xmax - xmin
        h = ymax - ymin
        x *= dw
        y *= dh
        w *= dw
        h *= dh
        return [x, y, w, h]

    def _get_annotations(self, annotation_path: str) -> dict:
        tree = ET.parse(annotation_path)
        root = tree.getroot()
        image_dir = root.find("folder").text
        filename = root.find("filename").text
        image_path = os.path.join(self.data_dir, image_dir, filename)
        size = root.find("size")
        width, height = int(size.find("width").text), int(size.find("height").text)
        bboxes, labels = [], []
        yolo_annotations = []

        for obj in root.findall("object"):
            label = obj.find("name").text
            label_id = self.labels_dict[label]
            bbox = obj.find("bndbox")
            xmin, ymin, xmax, ymax = (
                int(bbox.find(c).text) for c in ["xmin", "ymin", "xmax", "ymax"]
            )
            box = (xmin, ymin, xmax, ymax)
            bboxes.append(box)
            labels.append(label_id)

            xywh_box = self._convert_coordinates((width, height), box)
            xywh_box = " ".join([str(item) for item in xywh_box])

            yolo_annotations.append([str(label_id), xywh_box])

        return {
            "image_path": image_path,
            "bbox": bboxes,
            "labels": labels,
            "size": (width, height),
            "yolo_annotations": yolo_annotations,
        }

    def _save_annotations(self, annotations: list, filepath: str) -> None:
        with open(filepath, "w") as f:
            for annotation in annotations:
                line = " ".join([str(ann) for ann in annotation])
                f.write(line + "\n")

    def create_dataset_yolo(self, save_dir: str, config_file: str):
        train, val, test = self.split_dataset()
        train_annotations = [
            self._get_annotations(annotation_path)["yolo_annotations"]
            for annotation_path in train[1]
        ]

        val_annotations = [
            self._get_annotations(annotation_path)["yolo_annotations"]
            for annotation_path in val[1]
        ]
        test_annotations = [
            self._get_annotations(annotation_path)["yolo_annotations"]
            for annotation_path in test[1]
        ]

        def _save_dataset(data: tuple, split: str):
            dataset_dir = os.path.join(save_dir, split)
            os.makedirs(dataset_dir, exist_ok=True)
            images_dir = os.path.join(dataset_dir, "images")
            os.makedirs(images_dir, exist_ok=True)
            annotations_dir = os.path.join(dataset_dir, "labels")
            os.makedirs(annotations_dir, exist_ok=True)
            for i, (image_path, annotation) in enumerate(zip(data[0], data[1])):
                image_name = image_path.split("/")[-1]
                annotation_name = image_name[:-3] + "txt"
                shutil.copy(image_path, os.path.join(images_dir, image_name))
                print(f'Successfully created {os.path.join(images_dir, image_name)}')
                self._save_annotations(
                    annotation, os.path.join(annotations_dir, annotation_name)
                )
                print(f'Successfully created {os.path.join(annotations_dir, annotation_name)}')
                

        _save_dataset((train[0], train_annotations), "train")
        _save_dataset((val[0], val_annotations), "val")
        _save_dataset((test[0], test_annotations), "test")

        config = {
            "train": os.path.join(os.getcwd(), save_dir, "train", "images"),
            "val": os.path.join(os.getcwd(), save_dir, "val", "images"),
            "test": os.path.join(os.getcwd(), save_dir, "test", "images"),
            "nc": 3,
            "names": ["with_mask", "without_mask", "mask_weared_incorrect"],
        }

        with open(f"{save_dir}/{config_file}.yaml", "w") as yaml_file:
            yaml.dump(config, yaml_file, default_flow_style=False, sort_keys=False)
            print(f'Successfully created data config file {save_dir}/{config_file}.yaml')

    def split_dataset(self, split_size: float = 0.1, random_seed: int = 42) -> tuple:
        train_images, val_images, train_annotations, val_annotations = train_test_split(
            self.images_path, self.annotations_path, test_size=split_size, random_state=random_seed
        )
        train_images, test_images, train_annotations, test_annotations = train_test_split(
            train_images, train_annotations, test_size=split_size, random_state=random_seed
        )

        return (
            (train_images, train_annotations),
            (val_images, val_annotations),
            (test_images, test_annotations),
        )

## Face Mask Detection Using YOLOv8

### Import the library **Ultralytics**

In [3]:
from ultralytics import YOLO

### Generate and create dataset using class DatasetGenerator

In [4]:
DATASET_DIR = 'face_mask_dataset'

In [5]:
# dataset = DatasetGenerator(DATASET_DIR)
# dataset.create_dataset_yolo(save_dir='face_mask_dataset_yolo', config_file='data')

### Model Development

In [6]:
class YOLOTrainer:
    def __init__(self, model_path: str, data_path: str, project_name: str, hyp_path: str):
        self.model_path = model_path
        self.hyp_path = hyp_path
        self.data_path = data_path
        self.project_name = project_name
        self.hyperparameters = None
        self.model = None

    def load_model(self):
        self.model = YOLO(self.model_path)

    def load_hyperparameters(self):
        with open(self.hyp_path, 'r') as f:
            hyperparameters = yaml.safe_load(f)
        self.hyperparameters = hyperparameters

    def train(self, epochs: int, batch_size: int, optimizer: str = 'SGD', pretrained: bool = True):
        self.model.train(
            data=self.data_path,
            epochs=epochs,
            batch=batch_size,
            project=self.project_name,
            optimizer=optimizer,
            pretrained=pretrained,
            ** self.hyperparameters
        )

    def evaluate(self, validation_data: str):
        metrics = self.model.val(data=validation_data)
        return metrics

    def save_model(self, output_path: str):
        self.model.save(output_path)

### Model Training

In [7]:
trainer = YOLOTrainer(data_path='./yolo_format/data.yaml', model_path='yolov8n.pt', hyp_path='hyperparameters.yaml', project_name='face_mask_detection_yolov8')

In [8]:
trainer.load_model()

In [9]:
trainer.load_hyperparameters()

In [10]:
trainer.train(epochs=100, batch_size=64, optimizer='SGD')

New https://pypi.org/project/ultralytics/8.2.13 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.2.11 🚀 Python-3.10.4 torch-2.3.0+cu121 CUDA:0 (NVIDIA GeForce RTX 4080, 16079MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=./yolo_format/data.yaml, epochs=100, time=None, patience=100, batch=64, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=face_mask_detection_yolov8, name=train6, exist_ok=False, pretrained=True, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=15, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.5, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=Fals

/home/aidev6-pc/anaconda3/envs/face_mask_det.python3.10/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


AMP: checks passed ✅


/home/aidev6-pc/anaconda3/envs/face_mask_det.python3.10/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self.stride,
train: Scanning /home/aidev6-pc/dev/bimo/face_mask_detection/yolo_format/train/labels... 690 images, 0 backgrounds, 0 corrupt: 100%|██████████| 690/690 [00:00<00:00, 1808.37it/s]

train: New cache created: /home/aidev6-pc/dev/bimo/face_mask_detection/yolo_format/train/labels.cache


albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))


val: Scanning /home/aidev6-pc/dev/bimo/face_mask_detection/yolo_format/val/labels... 86 images, 0 backgrounds, 0 corrupt: 100%|██████████| 86/86 [00:00<00:00, 1359.33it/s]

val: New cache created: /home/aidev6-pc/dev/bimo/face_mask_detection/yolo_format/val/labels.cache


Plotting labels to face_mask_detection_yolov8/train6/labels.jpg... 


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited


optimizer: SGD(lr=0.01, momentum=0.937) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to face_mask_detection_yolov8/train6
Starting training for 100 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/11 [00:00<?, ?it/s]libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not 

                   all         86        535    0.00193      0.104    0.00139   0.000432



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
      2/100      10.2G      1.751      3.413      1.383        402        640: 100%|██████████| 11/11 [00:01<00:00,  6.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.99it/s]

                   all         86        535     0.0077      0.292      0.126     0.0667



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
      3/100      9.98G       1.46      1.961      1.191        635        640: 100%|██████████| 11/11 [00:01<00:00,  6.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.80it/s]

                   all         86        535     0.0103      0.458      0.243      0.149



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
      4/100      10.6G      1.437      1.637      1.132        466        640: 100%|██████████| 11/11 [00:01<00:00,  6.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.11it/s]


                   all         86        535     0.0137      0.645      0.251      0.156


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      5/100      11.6G      1.374      1.442      1.091        618        640: 100%|██████████| 11/11 [00:01<00:00,  6.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.18it/s]


                   all         86        535     0.0218      0.675      0.292      0.187

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
      6/100      9.23G      1.389      1.336      1.081        446        640: 100%|██████████| 11/11 [00:01<00:00,  6.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.30it/s]

                   all         86        535      0.024      0.689      0.322        0.2



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      7/100      10.7G       1.36      1.239      1.068        613        640: 100%|██████████| 11/11 [00:01<00:00,  5.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.30it/s]

                   all         86        535      0.991      0.109      0.318      0.191



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      8/100      10.7G      1.342      1.172      1.065        387        640: 100%|██████████| 11/11 [00:01<00:00,  5.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.29it/s]

                   all         86        535      0.562      0.142      0.405      0.235



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      9/100       9.4G      1.364      1.177      1.067        452        640: 100%|██████████| 11/11 [00:01<00:00,  5.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.27it/s]

                   all         86        535      0.953      0.253      0.468      0.286



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     10/100      9.95G      1.329      1.123      1.043        555        640: 100%|██████████| 11/11 [00:01<00:00,  6.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.14it/s]


                   all         86        535      0.603      0.424      0.503       0.29

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     11/100      10.6G      1.314      1.105      1.045        397        640: 100%|██████████| 11/11 [00:01<00:00,  6.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.32it/s]

                   all         86        535      0.525      0.388      0.445      0.266



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     12/100        11G      1.314      1.073      1.047        546        640: 100%|██████████| 11/11 [00:01<00:00,  6.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.26it/s]


                   all         86        535      0.577      0.454      0.506      0.288

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     13/100      10.3G      1.285      1.064      1.036        605        640: 100%|██████████| 11/11 [00:01<00:00,  6.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.42it/s]


                   all         86        535      0.832      0.359      0.427      0.241

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     14/100      10.5G      1.265     0.9921       1.04        469        640: 100%|██████████| 11/11 [00:01<00:00,  6.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.36it/s]

                   all         86        535      0.628      0.485        0.5      0.281



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     15/100      9.77G      1.253     0.9639      1.032        543        640: 100%|██████████| 11/11 [00:01<00:00,  6.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.85it/s]

                   all         86        535      0.843      0.476      0.513      0.285



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     16/100        11G      1.244     0.9626       1.02        447        640: 100%|██████████| 11/11 [00:01<00:00,  6.07it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.31it/s]


                   all         86        535      0.722      0.528      0.581       0.31

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     17/100      9.79G      1.245     0.9275      1.023        550        640: 100%|██████████| 11/11 [00:01<00:00,  6.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.26it/s]


                   all         86        535       0.84        0.5      0.609       0.37

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     18/100      9.33G      1.248     0.9231      1.025        503        640: 100%|██████████| 11/11 [00:01<00:00,  6.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.98it/s]

                   all         86        535      0.526      0.538      0.548      0.325



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     19/100      8.59G      1.218      0.892      1.018        520        640: 100%|██████████| 11/11 [00:01<00:00,  6.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.13it/s]


                   all         86        535      0.756      0.511      0.537      0.311


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     20/100      8.85G      1.255     0.9006      1.027        578        640: 100%|██████████| 11/11 [00:01<00:00,  6.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.36it/s]


                   all         86        535      0.666      0.568      0.607      0.319

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     21/100      10.8G      1.223     0.8954      1.019        531        640: 100%|██████████| 11/11 [00:01<00:00,  6.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.42it/s]

                   all         86        535       0.64      0.552      0.567      0.326



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     22/100       9.9G      1.198     0.8428      1.007        497        640: 100%|██████████| 11/11 [00:01<00:00,  6.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.67it/s]

                   all         86        535       0.67      0.656      0.689      0.411



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     23/100      9.97G      1.219      0.855      1.014        434        640: 100%|██████████| 11/11 [00:01<00:00,  6.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.69it/s]

                   all         86        535      0.579        0.6      0.593      0.372



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     24/100      10.4G      1.205     0.8194      1.001        527        640: 100%|██████████| 11/11 [00:01<00:00,  5.55it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.92it/s]

                   all         86        535      0.748      0.504      0.577      0.327



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     25/100      9.58G      1.183      0.818          1        390        640: 100%|██████████| 11/11 [00:02<00:00,  5.41it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.18it/s]

                   all         86        535        0.8      0.645      0.726      0.437



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     26/100         9G      1.175     0.8255      1.005        409        640: 100%|██████████| 11/11 [00:01<00:00,  6.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.61it/s]

                   all         86        535      0.713      0.667      0.699      0.429



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     27/100      10.6G      1.181     0.8011     0.9941        427        640: 100%|██████████| 11/11 [00:01<00:00,  6.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.49it/s]

                   all         86        535      0.869      0.661      0.723      0.448



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     28/100        11G      1.176     0.8147      1.002        671        640: 100%|██████████| 11/11 [00:01<00:00,  6.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.56it/s]


                   all         86        535      0.866      0.642      0.746      0.459


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     29/100      9.66G      1.181     0.8064     0.9973        472        640: 100%|██████████| 11/11 [00:01<00:00,  6.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.74it/s]

                   all         86        535      0.807      0.718      0.762      0.481



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     30/100      10.8G      1.183      0.788     0.9932        441        640: 100%|██████████| 11/11 [00:01<00:00,  6.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.64it/s]

                   all         86        535      0.906      0.632      0.739      0.462



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     31/100      10.1G      1.161     0.7677     0.9924        540        640: 100%|██████████| 11/11 [00:01<00:00,  6.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.33it/s]

                   all         86        535      0.772      0.653      0.716      0.452



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     32/100      10.4G      1.175     0.7778     0.9961        477        640: 100%|██████████| 11/11 [00:01<00:00,  6.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.62it/s]

                   all         86        535      0.876      0.619      0.733      0.437



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     33/100      9.77G      1.173     0.7662     0.9891        643        640: 100%|██████████| 11/11 [00:01<00:00,  6.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.69it/s]

                   all         86        535      0.817      0.666      0.752      0.455



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     34/100      9.92G      1.147     0.7615     0.9932        474        640: 100%|██████████| 11/11 [00:01<00:00,  6.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.67it/s]

                   all         86        535      0.797      0.662      0.712      0.422



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     35/100        10G      1.155     0.7566     0.9887        535        640: 100%|██████████| 11/11 [00:01<00:00,  6.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.58it/s]

                   all         86        535      0.886      0.587      0.725       0.44



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     36/100        11G      1.153     0.7483     0.9805        509        640: 100%|██████████| 11/11 [00:01<00:00,  6.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.56it/s]

                   all         86        535       0.91        0.7      0.768      0.462



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     37/100      9.99G      1.171     0.7487     0.9894        536        640: 100%|██████████| 11/11 [00:01<00:00,  6.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.75it/s]

                   all         86        535      0.906      0.672      0.783      0.492



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     38/100      10.2G      1.141     0.7449     0.9819        573        640: 100%|██████████| 11/11 [00:01<00:00,  6.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.76it/s]

                   all         86        535      0.915      0.635      0.785      0.474



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     39/100      10.9G      1.141      0.735     0.9878        347        640: 100%|██████████| 11/11 [00:01<00:00,  6.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.60it/s]

                   all         86        535       0.79      0.733      0.795       0.49



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     40/100      9.36G      1.147     0.7491     0.9822        626        640: 100%|██████████| 11/11 [00:01<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.60it/s]

                   all         86        535      0.792      0.731      0.768      0.457



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     41/100      9.81G      1.149      0.732     0.9814        615        640: 100%|██████████| 11/11 [00:01<00:00,  5.54it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.49it/s]

                   all         86        535       0.77      0.673      0.745      0.469



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     42/100      10.5G      1.138     0.7187     0.9851        494        640: 100%|██████████| 11/11 [00:01<00:00,  5.55it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.46it/s]

                   all         86        535      0.814      0.675      0.739      0.473



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     43/100      9.73G       1.13     0.7303      0.985        464        640: 100%|██████████| 11/11 [00:01<00:00,  6.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.66it/s]

                   all         86        535      0.755      0.697      0.742      0.456



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/11 [00:00<?, ?it/s]libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     44/100        10G      1.125     0.7055     0.9813        554        640: 100%|██████████| 11/11 [00:01<00:00,  6.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.67it/s]

                   all         86        535      0.891      0.641      0.746      0.453



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     45/100      9.83G      1.122      0.699      0.982        448        640: 100%|██████████| 11/11 [00:01<00:00,  6.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.61it/s]

                   all         86        535      0.921      0.665      0.773      0.491



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     46/100       9.7G      1.127     0.7143     0.9879        663        640: 100%|██████████| 11/11 [00:01<00:00,  6.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.60it/s]

                   all         86        535      0.841      0.656      0.771      0.481



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     47/100      8.94G      1.124     0.6816     0.9716        481        640: 100%|██████████| 11/11 [00:01<00:00,  6.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.70it/s]

                   all         86        535      0.832      0.679      0.757      0.477



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     48/100      10.5G      1.117     0.6991     0.9744        534        640: 100%|██████████| 11/11 [00:01<00:00,  6.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.65it/s]

                   all         86        535      0.874      0.601      0.729      0.481



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     49/100      10.3G      1.117     0.6993     0.9733        469        640: 100%|██████████| 11/11 [00:01<00:00,  6.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.73it/s]

                   all         86        535      0.791      0.687      0.739      0.469



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     50/100      8.87G      1.107     0.6889     0.9795        641        640: 100%|██████████| 11/11 [00:01<00:00,  6.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.76it/s]

                   all         86        535      0.866      0.615      0.715      0.438



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     51/100      9.93G      1.134     0.6956     0.9648        556        640: 100%|██████████| 11/11 [00:01<00:00,  6.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.48it/s]

                   all         86        535      0.789      0.628      0.733      0.463



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     52/100      9.51G      1.091     0.6739     0.9679        406        640: 100%|██████████| 11/11 [00:01<00:00,  6.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.60it/s]

                   all         86        535      0.711      0.694      0.743      0.493



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     53/100      9.38G      1.095     0.6757     0.9573        544        640: 100%|██████████| 11/11 [00:01<00:00,  6.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.64it/s]

                   all         86        535      0.875      0.662      0.772      0.498



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     54/100       8.9G      1.093     0.6566     0.9697        460        640: 100%|██████████| 11/11 [00:01<00:00,  6.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.66it/s]

                   all         86        535      0.837      0.714      0.778      0.502



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     55/100      9.73G      1.094     0.6599     0.9604        484        640: 100%|██████████| 11/11 [00:01<00:00,  6.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.15it/s]


                   all         86        535      0.882      0.709      0.776      0.516

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     56/100      9.94G      1.095     0.6558     0.9636        540        640: 100%|██████████| 11/11 [00:01<00:00,  6.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.68it/s]

                   all         86        535      0.794      0.718      0.784        0.5



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     57/100      10.2G      1.083     0.6465      0.961        429        640: 100%|██████████| 11/11 [00:01<00:00,  5.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.00it/s]

                   all         86        535      0.909      0.669      0.789      0.487



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     58/100      9.82G      1.106     0.6559     0.9635        476        640: 100%|██████████| 11/11 [00:02<00:00,  5.48it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.93it/s]

                   all         86        535      0.827      0.698      0.766      0.479



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     59/100      10.2G      1.088     0.6542     0.9631        428        640: 100%|██████████| 11/11 [00:01<00:00,  5.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.46it/s]

                   all         86        535      0.901      0.671       0.77      0.493



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     60/100      9.92G      1.092     0.6473     0.9599        549        640: 100%|██████████| 11/11 [00:01<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.63it/s]

                   all         86        535      0.881      0.747      0.807      0.518



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     61/100       9.7G      1.064      0.647     0.9584        395        640: 100%|██████████| 11/11 [00:01<00:00,  6.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.82it/s]

                   all         86        535      0.814      0.637      0.758        0.5



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     62/100      9.39G      1.062     0.6381     0.9566        671        640: 100%|██████████| 11/11 [00:01<00:00,  6.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.58it/s]

                   all         86        535      0.828      0.734      0.778      0.496



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     63/100      10.1G      1.064     0.6332     0.9503        415        640: 100%|██████████| 11/11 [00:01<00:00,  6.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.05it/s]

                   all         86        535      0.929      0.712      0.808      0.512



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     64/100      10.3G      1.042     0.6225     0.9468        639        640: 100%|██████████| 11/11 [00:01<00:00,  6.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.47it/s]

                   all         86        535      0.954      0.704      0.812      0.529



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     65/100      9.51G      1.042     0.6056     0.9435        702        640: 100%|██████████| 11/11 [00:01<00:00,  6.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.73it/s]

                   all         86        535      0.905      0.704      0.823      0.519



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     66/100      10.5G      1.065     0.6226     0.9487        505        640: 100%|██████████| 11/11 [00:01<00:00,  6.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.65it/s]

                   all         86        535      0.761      0.714      0.736      0.466



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     67/100      10.2G       1.05     0.6114     0.9463        508        640: 100%|██████████| 11/11 [00:01<00:00,  6.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.52it/s]

                   all         86        535      0.829       0.75      0.785      0.514



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     68/100      9.89G      1.061     0.6166     0.9503        577        640: 100%|██████████| 11/11 [00:01<00:00,  6.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.23it/s]

                   all         86        535      0.907      0.738      0.825      0.541



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     69/100       9.8G      1.045      0.611     0.9484        472        640: 100%|██████████| 11/11 [00:01<00:00,  6.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.51it/s]

                   all         86        535      0.858       0.71      0.794      0.521



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     70/100      10.5G      1.033     0.6014     0.9487        617        640: 100%|██████████| 11/11 [00:01<00:00,  6.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.42it/s]

                   all         86        535      0.834      0.708      0.777      0.507



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     71/100      10.7G      1.029     0.6056     0.9449        476        640: 100%|██████████| 11/11 [00:01<00:00,  6.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.59it/s]

                   all         86        535      0.798      0.739      0.764      0.497



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     72/100        10G      1.028     0.5904     0.9342        460        640: 100%|██████████| 11/11 [00:01<00:00,  6.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.57it/s]

                   all         86        535      0.849      0.739      0.791      0.517



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     73/100      11.1G      1.031     0.5923     0.9418        499        640: 100%|██████████| 11/11 [00:01<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.78it/s]

                   all         86        535      0.789      0.762      0.797      0.503



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     74/100      8.79G      1.032      0.609     0.9415        646        640: 100%|██████████| 11/11 [00:01<00:00,  5.54it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.78it/s]

                   all         86        535      0.831       0.76      0.789      0.517



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     75/100      9.92G      1.018     0.5858     0.9311        551        640: 100%|██████████| 11/11 [00:01<00:00,  5.55it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.59it/s]

                   all         86        535      0.817      0.795      0.809      0.519



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     76/100      10.1G      1.037      0.597      0.943        437        640: 100%|██████████| 11/11 [00:01<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.65it/s]

                   all         86        535      0.786      0.753      0.799      0.524



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     77/100      10.6G       1.03     0.5958     0.9421        551        640: 100%|██████████| 11/11 [00:01<00:00,  6.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.68it/s]

                   all         86        535      0.908      0.684      0.772      0.499



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     78/100      10.5G      1.034     0.5959      0.942        631        640: 100%|██████████| 11/11 [00:01<00:00,  6.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.72it/s]

                   all         86        535      0.872      0.741      0.818      0.516



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/11 [00:00<?, ?it/s]libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     79/100      11.4G      1.007     0.5851     0.9302        565        640: 100%|██████████| 11/11 [00:01<00:00,  6.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.05it/s]


                   all         86        535       0.85      0.756      0.802      0.521

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     80/100        12G      1.022     0.5812     0.9373        452        640: 100%|██████████| 11/11 [00:01<00:00,  6.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.52it/s]

                   all         86        535      0.893      0.729       0.81      0.531



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/11 [00:00<?, ?it/s]libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     81/100      10.4G      1.014     0.5879      0.944        616        640: 100%|██████████| 11/11 [00:01<00:00,  6.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.72it/s]

                   all         86        535      0.896      0.724      0.817      0.535



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     82/100      10.2G     0.9914     0.5748     0.9275        352        640: 100%|██████████| 11/11 [00:01<00:00,  6.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.76it/s]

                   all         86        535      0.923      0.669      0.813      0.531



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     83/100      11.3G      1.013     0.5812     0.9391        427        640: 100%|██████████| 11/11 [00:01<00:00,  6.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.70it/s]

                   all         86        535      0.909        0.7      0.819      0.528



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     84/100      9.09G     0.9986     0.5689     0.9365        461        640: 100%|██████████| 11/11 [00:01<00:00,  6.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.81it/s]

                   all         86        535      0.866      0.718      0.813      0.529



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     85/100      10.5G      1.001     0.5617     0.9366        512        640: 100%|██████████| 11/11 [00:01<00:00,  6.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.76it/s]

                   all         86        535      0.905      0.685      0.789      0.522


Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     86/100      10.1G     0.9377     0.5224     0.9318        201        640: 100%|██████████| 11/11 [00:02<00:00,  3.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.43it/s]

                   all         86        535      0.826      0.686      0.776      0.523



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/11 [00:00<?, ?it/s]libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     87/100        10G      0.936     0.4884     0.9246        162        640: 100%|██████████| 11/11 [00:01<00:00,  6.46it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.05it/s]

                   all         86        535      0.882      0.704      0.805      0.535



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/11 [00:00<?, ?it/s]libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     88/100      9.74G     0.9129     0.4856     0.9157        259        640: 100%|██████████| 11/11 [00:01<00:00,  6.43it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.72it/s]

                   all         86        535      0.884      0.706      0.805      0.524



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/11 [00:00<?, ?it/s]libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     89/100      10.2G     0.9163     0.4919     0.9208        236        640: 100%|██████████| 11/11 [00:01<00:00,  6.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.21it/s]

                   all         86        535      0.891      0.724      0.808      0.528



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     90/100      9.43G     0.9059     0.4747     0.9165        222        640: 100%|██████████| 11/11 [00:01<00:00,  5.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.08it/s]

                   all         86        535      0.866       0.75      0.817      0.525



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     91/100      9.27G     0.9245     0.4832     0.9272        181        640: 100%|██████████| 11/11 [00:01<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.30it/s]

                   all         86        535       0.86      0.764      0.838      0.552



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     92/100      9.75G     0.9034     0.4681     0.9117        226        640: 100%|██████████| 11/11 [00:01<00:00,  6.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.60it/s]

                   all         86        535      0.916      0.732      0.803      0.519



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     93/100      9.28G     0.8883     0.4661     0.9154        269        640: 100%|██████████| 11/11 [00:01<00:00,  6.47it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.62it/s]

                   all         86        535      0.922      0.719      0.807      0.529



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     94/100      9.27G     0.9004     0.4612     0.9166        276        640: 100%|██████████| 11/11 [00:01<00:00,  6.50it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.86it/s]

                   all         86        535      0.921      0.762      0.819      0.536



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     95/100      9.76G     0.8939     0.4567     0.9179        206        640: 100%|██████████| 11/11 [00:01<00:00,  6.54it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.79it/s]

                   all         86        535      0.915      0.723      0.813      0.526



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     96/100      10.2G     0.8785     0.4539     0.9134        230        640: 100%|██████████| 11/11 [00:01<00:00,  6.49it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.73it/s]

                   all         86        535      0.909      0.777      0.828      0.534



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     97/100       9.9G     0.8884     0.4505     0.9102        191        640: 100%|██████████| 11/11 [00:01<00:00,  6.40it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.72it/s]

                   all         86        535      0.914      0.736      0.823      0.536



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
     98/100      9.72G     0.8781     0.4453     0.9094        178        640: 100%|██████████| 11/11 [00:01<00:00,  6.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.78it/s]

                   all         86        535      0.855      0.758      0.819      0.532



libpng warning: iCCP: Not recognizing known sRGB profile that has been edited



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     99/100      9.74G     0.8598     0.4396     0.9075        226        640: 100%|██████████| 11/11 [00:01<00:00,  6.45it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.74it/s]

                   all         86        535      0.916      0.726      0.811      0.529



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
    100/100      10.3G     0.8699     0.4524     0.9081        255        640: 100%|██████████| 11/11 [00:01<00:00,  6.45it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  5.73it/s]

                   all         86        535      0.903      0.772       0.82       0.53



100 epochs completed in 0.067 hours.
Optimizer stripped from face_mask_detection_yolov8/train6/weights/last.pt, 6.3MB
Optimizer stripped from face_mask_detection_yolov8/train6/weights/best.pt, 6.3MB

Validating face_mask_detection_yolov8/train6/weights/best.pt...
Ultralytics YOLOv8.2.11 🚀 Python-3.10.4 torch-2.3.0+cu121 CUDA:0 (NVIDIA GeForce RTX 4080, 16079MiB)


libpng warning: iCCP: Not recognizing known sRGB profile that has been edited


Model summary (fused): 168 layers, 3006233 parameters, 0 gradients, 8.1 GFLOPs


/home/aidev6-pc/anaconda3/envs/face_mask_det.python3.10/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self.stride,
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  6.28it/s]
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited


                   all         86        535       0.86      0.764      0.838      0.553
             with_mask         86        464      0.932      0.914      0.944      0.641
          without_mask         86         55       0.75      0.818      0.857      0.521
 mask_weared_incorrect         86         16      0.899      0.559      0.713      0.496
Speed: 0.1ms preprocess, 0.5ms inference, 0.0ms loss, 0.2ms postprocess per image
Results saved to face_mask_detection_yolov8/train6


### Model Evaluation

In [12]:
metrics = trainer.evaluate(validation_data='./yolo_format/data.yaml')

Ultralytics YOLOv8.2.11 🚀 Python-3.10.4 torch-2.3.0+cu121 CUDA:0 (NVIDIA GeForce RTX 4080, 16079MiB)


val: Scanning /home/aidev6-pc/dev/bimo/face_mask_detection/yolo_format/val/labels.cache... 86 images, 0 backgrounds, 0 corrupt: 100%|██████████| 86/86 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/2 [00:00<?, ?it/s]libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
/home/aidev6-pc/anaconda3/envs/face_mask_det.python3.10/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self

                   all         86        535      0.862      0.764      0.834      0.547
             with_mask         86        464      0.934      0.916      0.944       0.64
          without_mask         86         55      0.752      0.818      0.854      0.516
 mask_weared_incorrect         86         16      0.899      0.558      0.705      0.485
Speed: 0.1ms preprocess, 1.6ms inference, 0.0ms loss, 0.8ms postprocess per image
Results saved to face_mask_detection_yolov8/train63


If we see the result of model evaluation, we got some metrics.
1. **Box Precision (P)**
- The precision metric for the bounding box predictions. It measures the accuracy of the detected boxes in terms of how many detected boxes are actually correct. Then we got, a precision of 0.934 for "with_mask" means that 93.4% of the boxes around faces with masks are correctly placed.
![Precision(P)](face_mask_detection_yolov8/train63/P_curve.png)

2. **Recall (R)**
- This metric measures the model's ability to find all the relevant cases (instances) within the datasets. For "without_mask", a recall of 0.818 means the model correctly identifies 81.8% of all actual "without_mask" instances in the test set.
![Recall(R)](face_mask_detection_yolov8/train63/R_curve.png)

3. **mAP50**
- Mean Average Precision at 50% IoU (Intersection over Union). This is a common metric for evaluating object detection models, where a detection is considered correct if the IoU is above 50%. The mAP50 of 0.944 for "with_mask" is particularly high, indicating excellent model performance for this class at this IoU threshold.

4. **mAP50-95**
- This is an average of the mAP calculated at different IoU thresholds from 50% to 95% (inclusive), in steps of 5%. It is a more stringent measure since it considers a range of IoU thresholds. The value of 0.64 for "with_mask" under this metric shows the model performs well even under stricter criteria, though naturally, the score is lower than the mAP50. You can see the results, in this figure below:
![Results](face_mask_detection_yolov8/train6/results.png)

### Conclusion

1. **Overall Performance**:
The model performs very well on detecting faces with masks, with high precision and mAP scores, which because that the training data likely had a large number of "with_mask" examples, or the model features are particularly good at picking out this class.

2. **Model Robustness**: The model is generally robust, as indicated by consistent training and validation loss curves, and maintains good precision across high confidence levels.

3. The model shows excellent precision and recall for faces with masks.

4. The model achieves commendable mAP50 and mAP50-95 scores, especially in the "with_mask" category, 

5. The model struggle slightly with detecting "mask_weared_incorrect", as indicated by the lower recall and mAP scores.



### WHAT SHOULD WE DO?

1. Add more images that represent diverse scenarios, especially focusing on underrepresented classes like "without_mask" and "mask_weared_incorrect". Make sure to include a variety of backgrounds, lighting conditions, and face orientations.

2. Apply transformations such as rotations, scaling, cropping, and color adjustments to existing images to create a more robust training dataset. This helps the model generalize better across different real-world conditions.

3. Adjust Hyperparameters: Optimize learning rates, decay schedules, and other hyperparameters. Sometimes, small adjustments can yield significant improvements.

## Face Mask Detection (using Classification)

### Import libraries needed

In [ ]:
import torch
from transformers import (
    ViTImageProcessor,
    ViTModel,
    ViTForImageClassification,
    Trainer,
    TrainingArguments
)

from torchvision.transforms import (
    Compose,
    Normalize,
    RandomResizedCrop,
    RandomHorizontalFlip,
    CenterCrop,
    Resize,
    ToTensor,
)

from sklearn.utils.class_weight import compute_sample_weight
from datasets import Dataset, DatasetDict, load_metric, ClassLabel


#### Initialize parameters

In [ ]:
DATA_DIR = "./face_mask_dataset/"
MODEL_CHECKPOINT = "google/vit-base-patch16-224-in21k"
DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
MODEL_OUTPUT = "vit-face-mask-classification"
NUM_OF_EPOCHS = 10
BATCH_SIZE = 64
STRATEGY = "epoch"
LR = 5e-5

### Generate Data from DatasetGenerator

In [ ]:
DATA_DIR = './face_mask_dataset'

In [ ]:
dataset = DatasetGenerator(DATA_DIR)

In [ ]:
dataset.to_dict()

{'image': [<PIL.Image.Image image mode=RGB size=30x37>,
  <PIL.Image.Image image mode=RGB size=41x44>,
  <PIL.Image.Image image mode=RGB size=35x51>,
  <PIL.Image.Image image mode=RGB size=33x35>,
  <PIL.Image.Image image mode=RGB size=37x35>,
  <PIL.Image.Image image mode=RGB size=16x23>,
  <PIL.Image.Image image mode=RGB size=31x41>,
  <PIL.Image.Image image mode=RGB size=21x30>,
  <PIL.Image.Image image mode=RGB size=30x26>,
  <PIL.Image.Image image mode=RGB size=23x20>,
  <PIL.Image.Image image mode=RGB size=29x29>,
  <PIL.Image.Image image mode=RGB size=28x33>,
  <PIL.Image.Image image mode=RGB size=96x116>,
  <PIL.Image.Image image mode=RGB size=56x58>,
  <PIL.Image.Image image mode=RGB size=13x21>,
  <PIL.Image.Image image mode=RGB size=45x39>,
  <PIL.Image.Image image mode=RGB size=116x106>,
  <PIL.Image.Image image mode=RGB size=39x49>,
  <PIL.Image.Image image mode=RGB size=52x56>,
  <PIL.Image.Image image mode=RGB size=48x60>,
  <PIL.Image.Image image mode=RGB size=42x42>,
 

### Preprocessing Data for Training Model

In [ ]:
id2label = dataset.labels_dict
label2id = {v: k for k, v in id2label.items()}

In [ ]:
ds = Dataset.from_dict(dataset.to_dict())
ds = ds.cast_column("labels", ClassLabel(names=list(id2label.keys())))

Casting the dataset: 100%|██████████| 4072/4072 [00:00<00:00, 268532.53 examples/s]


#### Splitting Dataset

In [ ]:
# Splitting dataset into train and val
ds = ds.train_test_split(test_size=0.1, shuffle=True, seed=42)
ds = DatasetDict({"train": ds["train"], "val": ds["test"]})

#### Transformation Dataset

In [ ]:
def preprocess_image(example_batch, transforms):
    example_batch["pixel_values"] = [
        transforms(image.convert("RGB")) for image in example_batch["image"]
    ]
    return example_batch

In [ ]:
processor = ViTImageProcessor.from_pretrained(MODEL_CHECKPOINT)

In [ ]:
image_transforms = Compose(
    [
        Resize((processor.size["height"], processor.size["width"])),
        RandomHorizontalFlip(),
        ToTensor(),
        Normalize(
            mean=processor.image_mean,
            std=processor.image_std,
        ),
    ]
)

In [ ]:
# Applied transform
ds["train"].set_transform(lambda x: preprocess_image(x, image_transforms))
ds["val"].set_transform(lambda x: preprocess_image(x, image_transforms))

### Training Model

In [ ]:
def compute_metrics(p):
    accuracy_metric = load_metric("accuracy")
    accuracy = accuracy_metric.compute(
        predictions=np.argmax(p.predictions, axis=1), references=p.label_ids
    )["accuracy"]
    f1_score_metric = load_metric("f1")
    weighted_f1_score = f1_score_metric.compute(
        predictions=np.argmax(p.predictions, axis=1),
        references=p.label_ids,
        average="weighted",
    )["f1"]
    recall_metric = load_metric("recall")
    weighted_recall = recall_metric.compute(
        predictions=np.argmax(p.predictions, axis=1),
        references=p.label_ids,
        average="weighted",
    )["recall"]
    precision_metric = load_metric("precision")
    weighted_precision = precision_metric.compute(
        predictions=np.argmax(p.predictions, axis=1),
        references=p.label_ids,
        average="weighted",
    )["precision"]
    return {
        "accuracy": accuracy,
        "Weighted F1": weighted_f1_score,
        "Weighted Recall": weighted_recall,
        "Weighted Precision": weighted_precision,
    }

In [ ]:
def collate_fn(example_batch):
    pixel_values = torch.stack([example["pixel_values"] for example in example_batch])
    labels = torch.tensor([example["labels"] for example in example_batch])
    return {"pixel_values": pixel_values, "labels": labels}

In [ ]:
model = ViTForImageClassification.from_pretrained(
    MODEL_CHECKPOINT,
    num_labels=len(id2label),
    id2label=id2label,
    label2id=label2id,
)
model.to(DEVICE)

Some weights of ViTForImageClassification were not initialized from the model checkpoint at google/vit-base-patch16-224-in21k and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


ViTForImageClassification(
  (vit): ViTModel(
    (embeddings): ViTEmbeddings(
      (patch_embeddings): ViTPatchEmbeddings(
        (projection): Conv2d(3, 768, kernel_size=(16, 16), stride=(16, 16))
      )
      (dropout): Dropout(p=0.0, inplace=False)
    )
    (encoder): ViTEncoder(
      (layer): ModuleList(
        (0-11): 12 x ViTLayer(
          (attention): ViTAttention(
            (attention): ViTSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.0, inplace=False)
            )
            (output): ViTSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.0, inplace=False)
            )
          )
          (intermediate): ViTIntermediate(
            (dense): Linear(in_features=7

In [ ]:
args = TrainingArguments(
    MODEL_OUTPUT,
    remove_unused_columns=False,
    evaluation_strategy=STRATEGY,
    save_strategy=STRATEGY,
    logging_strategy="steps",
    logging_steps=10,
    logging_first_step=True,
    learning_rate=LR,
    per_device_train_batch_size=BATCH_SIZE,
    per_device_eval_batch_size=BATCH_SIZE,
    gradient_accumulation_steps=4,
    num_train_epochs=NUM_OF_EPOCHS,
    warmup_ratio=0.10,
    save_total_limit=1,
)

In [ ]:
trainer = Trainer(
    model=model,
    args=args,
    train_dataset=ds["train"],
    eval_dataset=ds["val"],
    tokenizer=processor,
    data_collator=collate_fn,
    compute_metrics=compute_metrics,
)

In [ ]:
train_results = trainer.train()

In [ ]:
trainer.save_model()

In [ ]:
trainer.log_metrics("train", train_results.metrics)
trainer.save_metrics("train", train_results.metrics)
trainer.save_state()

eval_results = trainer.evaluate(ds["val"])
trainer.log_metrics("eval", eval_results.metrics)
trainer.save_metrics("eval", eval_results.metrics)
trainer.save_state()

### Model Evaluation

#### Import libraries needed

In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.metrics import roc_curve, auc
import matplotlib.pyplot as plt
import seaborn as sns

### Load Trained Model

In [ ]:
model = ViTForImageClassification.from_pretrained(MODEL_OUTPUT)
processor = ViTImageProcessor.from_pretrained(MODEL_OUTPUT)

### Generate Data for Evaluation

In [ ]:
dataset = DatasetGenerator(DATASET_DIR)

In [ ]:
results = []

data = {
    "ground_truth": [],
    "label": [],
    "label_name": [],
    "probs": [],
}

for image, label in dataset:
    ground_truth = label
    inputs = processor(images=image, return_tensors="pt")

    with torch.no_grad():
        outputs = model(**inputs)
        logits = outputs.logits
        probabilities = torch.nn.functional.softmax(logits, dim=-1)

    predicted_class_idx = probabilities.argmax(-1).item()
    predicted_class = model.config.id2label[predicted_class_idx]

    results.append(predicted_class_idx)

    data["ground_truth"].append(ground_truth)
    data["label"].append(predicted_class_idx)
    data["label_name"].append(predicted_class)
    data["probs"].append(probabilities.max().detach().cpu().numpy())

In [ ]:
df = pd.DataFrame(data)
df.head()

ground_truth  label    label_name       probs
0             1      1  without_mask  0.94091743
1             0      0     with_mask  0.96266997
2             1      1  without_mask   0.9440221
3             0      0     with_mask   0.9663677
4             0      0     with_mask   0.9665936

In [ ]:
df.to_csv('results.csv', index=True)

In [ ]:
y_true = df["ground_truth"].values
y_pred = df["label"].values

In [ ]:
accuracy = accuracy_score(y_true, y_pred)
print(f"Accuracy: {accuracy:.4f}")
# print(classification_report(y_true, y_pred))
# print(confusion_matrix(y_true, y_pred))

Accuracy: 0.9912


#### Visualize

In [ ]:
def plot_confusion_matrix(y_true, y_pred):
    cm = confusion_matrix(y_true, y_pred)
    plt.figure(figsize=(8, 6))
    sns.heatmap(cm, annot=True, fmt="d", cmap="Blues", cbar=False)
    plt.xlabel("Predicted")
    plt.ylabel("True")
    plt.title("Confusion Matrix")
    plt.show()

In [ ]:
plot_confusion_matrix(y_true, y_pred)

<Figure size 800x600 with 1 Axes>